### ISSUE TITLE
GetHash() issues

### ISSUE THREAD

**Issue by virzak** [Ne4to/N.SourceGenerators.UnionTypes#11]
There are 2 issues with GetHash()

- It uses boxing/unboxing and can be improved significantly. See #10
- Should it include the type?

```cs
public override int GetHashCode()
{
    if (IsSuccessStruct)
        return HashCode.Combine(_variant, AsSuccessStruct);
    if (IsValidationErrorStruct)
        return HashCode.Combine(_variant, AsValidationErrorStruct);
    if (IsNotFoundErrorStruct)
        return HashCode.Combine(_variant, AsNotFoundErrorStruct);
    throw new InvalidOperationException("Unknown type");
}
```

**Comment by Ne4to** [comment]
Make sense, I'll implement it later. `ToString` could be optimized the same way.
I don't think type is required. It's unlikely to have hash code collision from distinct types, I guess.

**Comment by Ne4to** [comment]
fixed in `v0.16.0`
